src/threading/TaskScheduler.cpp

```diff
@@ -18,62 +19,231 @@
+        }
+
+        // Make sure the task queue is created for this pool
+        /* mutex scope */ {
+            const std::lock_guard<std::mutex> queue_lock(queue_mutex);
+            if (queue.count(pool.pool_id) == 0) {
+                queue.emplace(pool.pool_id, std::vector<std::unique_ptr<ReactionTask>>{});
+            }
+        }
+
+        // If the scheduler has not yet started then don't start the threads for this pool yet
+        if (started.load()) {
+            start_threads(pool);
         }
-        condition.notify_all();
+    }
+
+    void TaskScheduler::start(const size_t& thread_count) {
+
+        // Make the default pool
+        create_pool(util::ThreadPoolDescriptor{util::ThreadPoolDescriptor::DEFAULT_THREAD_POOL_ID, thread_count});
+
+        // The scheduler is now started
+        started.store(true);
+
+        // Start all our threads
+        for (const auto& pool : pools) {
+            start_threads(pool.second);
+        }
+
+        // Run main thread tasks
+        pool_func(pools.at(util::ThreadPoolDescriptor::MAIN_THREAD_POOL_ID));
+
+        // Poke all of the threads to make sure they are awake
+        /* mutex scope */ {
+            const std::lock_guard<std::mutex> queue_lock(queue_mutex);
+            queue_condition.notify_all();
+        }
+
+        // Now wait for all the threads to finish executing
+        for (auto& thread : threads) {
+            try {
+                if (thread->joinable()) {
+                    thread->join();
+                }
+            }
+            // This gets thrown some time if between checking if joinable and joining
+            // the thread is no longer joinable
+            catch (const std::system_error&) {
+            }
+        }
+    }
+
+    void TaskScheduler::shutdown() {
+        started.store(false);
+        running.store(false);
+        const std::lock_guard<std::mutex> queue_lock(queue_mutex);
+        queue_condition.notify_all();
     }
 
     void TaskScheduler::submit(std::unique_ptr<ReactionTask>&& task) {
 
-        // We do not accept new tasks once we are shutdown
-        if (running) {
+        // Extract the thread pool descriptor from the current task
+        const util::ThreadPoolDescriptor current_pool = task->thread_pool_descriptor;
+
+        // Make sure the pool is created
+        create_pool(current_pool);
+
+        // Make sure we know about this group
+        /* mutex scope */ {
+            const std::lock_guard<std::mutex> group_lock(queue_mutex);
+            const uint64_t group_id = task->group_descriptor.group_id;
+            if (groups.count(group_id) == 0) {
+                groups.emplace(group_id, 0);
+            }
+        }
+
+        // Check to see if this task was the result of `emit<Direct>`
+        // Direct tasks can run after shutdown and before starting, provided they can be run immediately
+        if (task->immediate) {
```

Comment:
I think it's a little trickier here, do we want to only care about group and not pool?
E.g. if a MainThread task emits direct we want it to run.
That happens a lot on startup e.g. on<Configuration> extensions do an immediate execution during their construction which is all on the main thread. For those to still work pool needs to be ignored when immediate is true.

We should also make a list of all these things as we define them so we can make unit tests for them

Reply:
Yea Im not sure what we want to do with immediate tasks. All of the current tests pass (log tests are probably more important for this), but we also dont have much of a test for an extension and how that would work, except for IOController and networking (the networking test also seems to work)